File: libs/solvers/unittests/test_adapt_mpi.cpp
/*******************************************************************************
 * Copyright (c) 2024 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// ADAPT-VQE MPI work-splitting test.
// Verifies that ADAPT-VQE produces the correct H2 ground-state energy when
// commutator evaluation is distributed across multiple MPI ranks.
//
// When run by gtest (ctest), the test launches itself via mpiexec as a
// subprocess. When invoked with --mpi-worker, the actual MPI computation runs.

#include <cmath>
#include <cstdlib>
#include <iostream>
#include <string>

#include <gtest/gtest.h>

#include "cudaq.h"
#include "nvqpp/test_kernels.h"
#include "cudaq/solvers/adapt.h"
#include "cudaq/solvers/operators.h"

// Lightweight MPI init/finalize used by the skip-check probe.
static int runMpiProbe() {
  cudaq::mpi::initialize();
  cudaq::mpi::finalize();
  return 0;
}

static int runMpiWorker() {
  cudaq::mpi::initialize();

  auto geometryHH = cudaq::solvers::molecular_geometry{{"H", {0., 0., 0.}},
                                                       {"H", {0., 0., .7474}}};
  auto hh = cudaq::solvers::create_molecule(
      geometryHH, "sto-3g", 0, 0,
      {.casci = true, .ccsd = false, .verbose = false});

  auto pool = cudaq::solvers::operator_pool::get("spin_complement_gsd");
  auto poolList =
      pool->generate({{"num-orbitals", hh.hamiltonian.num_qubits() / 2}});

  auto [energy, thetas, ops] =
      cudaq::solvers::adapt_vqe(hartreeFock2Electrons, hh.hamiltonian, poolList,
                                {{"grad_norm_tolerance", 1e-3},
                                 {"max_iter", 15},
                                 {"grad_norm_diff_tolerance", 1e-5},
                                 {"threshold_energy", 5e-6}});

  int rc = 0;
  if (cudaq::mpi::rank() == 0) {
    std::cout << "[MPI ADAPT] ranks=" << cudaq::mpi::num_ranks()
              << ", energy=" << energy << std::endl;
    if (std::fabs(energy - (-1.13)) >= 1e-2) {
      std::cerr << "FAIL: energy " << energy << " != expected -1.13"
                << std::endl;
      rc = 1;
    } else if (ops.empty()) {
      std::cerr << "FAIL: no operators selected" << std::endl;
      rc = 1;
    } else {
      std::cout << "PASS" << std::endl;
    }
  }

  cudaq::mpi::finalize();
  return rc;
}

class SolversTester : public ::testing::TestWithParam<int> {};

TEST_P(SolversTester, checkSimpleAdaptMpi_H2Sto3g) {
  if (std::system("which mpiexec > /dev/null 2>&1") != 0)
    GTEST_SKIP() << "mpiexec not found, skipping MPI test";

  // Probe with 2 ranks to verify MPI can actually launch multi-rank jobs
  // (catches missing PML transports, absent cudaq MPI plugin, etc.)
  std::string self = ::testing::internal::GetArgvs()[0];
  std::string probeCmd = "mpiexec --allow-run-as-root --oversubscribe -np 2 " +
                         self + " --mpi-probe > /dev/null 2>&1";
  if (std::system(probeCmd.c_str()) != 0)
    GTEST_SKIP() << "MPI multi-rank launch not functional, skipping MPI test";

  int numRanks = GetParam();

  std::string cmd = "mpiexec --allow-run-as-root --oversubscribe -np " +
                    std::to_string(numRanks) + " " + self + " --mpi-worker";
  int rc = std::system(cmd.c_str());
  EXPECT_EQ(rc, 0) << "mpiexec failed with exit code " << rc;
}

INSTANTIATE_TEST_SUITE_P(MpiRanks, SolversTester, ::testing::Values(2, 4));

int main(int argc, char **argv) {
  for (int i = 1; i < argc; i++) {
    if (std::string(argv[i]) == "--mpi-probe")
      return runMpiProbe();
    if (std::string(argv[i]) == "--mpi-worker")
      return runMpiWorker();
  }

  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: libs/solvers/python/tests/test_adapt_mpi.py
# ============================================================================ #
# Copyright (c) 2024 - 2026 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

# ADAPT-VQE MPI work-splitting test.
# Verifies that ADAPT-VQE produces the correct H2 ground-state energy when
# commutator evaluation is distributed across multiple MPI ranks.
#
# When invoked by pytest, the test launches itself via mpiexec as a subprocess.
# When invoked by mpiexec, the __name__ == "__main__" block runs the actual
# MPI computation.

import os
import shutil
import subprocess

import pytest

_MPI_CMD = ["mpiexec", "--allow-run-as-root", "--oversubscribe"]


def _mpi_available():
    if shutil.which("mpiexec") is None:
        return False
    # Probe with 2 ranks to verify multi-rank MPI actually works
    # (catches missing PML transports, absent cudaq MPI plugin, etc.)
    try:
        rc = subprocess.run(_MPI_CMD + [
            "-np", "2", "python3", "-c",
            "import cudaq; cudaq.mpi.initialize(); "
            "cudaq.mpi.finalize()"
        ],
                            capture_output=True,
                            timeout=30)
        return rc.returncode == 0
    except subprocess.TimeoutExpired:
        return False


@pytest.mark.skipif(not _mpi_available(),
                    reason="mpiexec or cudaq MPI support not available")
@pytest.mark.parametrize("num_ranks", [2, 4])
def test_adapt_mpi(num_ranks):
    result = subprocess.run(_MPI_CMD +
                            ["-np", str(num_ranks), "python3", __file__],
                            capture_output=True,
                            text=True,
                            timeout=120)
    assert result.returncode == 0, \
        (f"MPI test failed (np={num_ranks}):\n"
         f"--- stdout ---\n{result.stdout}\n"
         f"--- stderr ---\n{result.stderr}")


if __name__ == "__main__":
    import numpy as np
    import cudaq
    import cudaq_solvers as solvers

    cudaq.mpi.initialize()

    geometry = [('H', (0., 0., 0.)), ('H', (0., 0., .7474))]
    molecule = solvers.create_molecule(geometry, 'sto-3g', 0, 0, casci=True)
    operators = solvers.get_operator_pool("spin_complement_gsd",
                                          num_orbitals=molecule.n_orbitals)
    numElectrons = molecule.n_electrons

    @cudaq.kernel
    def initState(q: cudaq.qview):
        for i in range(numElectrons):
            x(q[i])

    energy, thetas, ops = solvers.adapt_vqe(initState, molecule.hamiltonian,
                                            operators)

    rank = cudaq.mpi.rank()
    num_ranks = cudaq.mpi.num_ranks()

    if rank == 0:
        print(f"[MPI ADAPT] ranks={num_ranks}, energy={energy:.6f}")
        assert np.isclose(energy, -1.137, atol=1e-3), \
            f"MPI ADAPT energy {energy} does not match expected -1.137"
        assert len(ops) > 0, "Expected at least one operator to be selected"
        print("PASS")

    cudaq.mpi.finalize()
